### DISCUSSION TITLE
syncRejectBlocklistedTorrentHashesWhileGrabbing

### DISCUSSION THREAD

**Discussion by ezekieldas** [buildarr/buildarr#206] [Q&A] [answered]
note: I'm making a few comments below as this evolves... 

I was about to wrap up the first segment of my project in using buildarr for full, first-time provisioning and this issue resurfaced. I'm having a hard time determining what's causing it or where it's coming from (although still investigating). 

My buildarr.yml looks like this so I can easily include/exclude indexers and apps. I'm going to try provisioning the stack without indexers/apps first and see if that has posiitive impact. 

```
includes:
  - prowlarr.yml
  - prowlarr.indexers.yml
  - prowlarr.apps.yml
  - radarr.yml
  - sonarr.yml
```

https://github.com/search?q=repo%3AProwlarr%2FProwlarr%20syncRejectBlocklistedTorrentHashesWhileGrabbing&type=code

https://github.com/buildarr/buildarr-prowlarr/blob/main/buildarr_prowlarr/config/settings/apps/applications.py#L213

```
2024-04-27 15:48:06,550 buildarr:65 buildarr.cli.run [INFO] <prowlarr> (default) Fetching remote configuration to check if updates are required
2024-04-27 15:48:06,660 buildarr:65 buildarr.cli.run [INFO] <prowlarr> (default) Finished fetching remote configuration
2024-04-27 15:48:06,684 buildarr:65 buildarr.cli.run [INFO] <prowlarr> (default) Updating remote configuration
2024-04-27 15:48:06,721 buildarr:65 buildarr.config.base [INFO] <prowlarr> (default) prowlarr.settings.apps.applications.definitions['Sonarr'].api_key: '**********' -> '**********'
Traceback (most recent call last):
  File "/usr/local/bin/buildarr", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/buildarr/cli/run.py", line 101, in run
    _run(use_plugins)
  File "/usr/local/lib/python3.11/site-packages/buildarr/cli/run.py", line 248, in _run
    if manager.update_remote(
       ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/buildarr/manager/__init__.py", line 192, in update_remote
    return local_instance_config.update_remote(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/buildarr/config/base.py", line 272, in update_remote
    if isinstance(field, ConfigBase) and field.update_remote(
                                         ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/buildarr_prowlarr/config/settings/__init__.py", line 70, in update_remote
    self.apps.update_remote(
  File "/usr/local/lib/python3.11/site-packages/buildarr/config/base.py", line 272, in update_remote
    if isinstance(field, ConfigBase) and field.update_remote(
                                         ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/buildarr_prowlarr/config/settings/apps/applications.py", line 748, in update_remote
    elif local_application._update_remote(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/buildarr_prowlarr/config/settings/apps/applications.py", line 233, in _update_remote
    set_attrs["fields"] = [
                          ^
  File "/usr/local/lib/python3.11/site-packages/buildarr_prowlarr/config/settings/apps/applications.py", line 234, in <listcomp>
    {**f, "value": field_values[f["name"]]} for f in api_schema.to_dict()["fields"]
                   ~~~~~~~~~~~~^^^^^^^^^^^
KeyError: 'syncRejectBlocklistedTorrentHashesWhileGrabbing'
```

**Comment by ezekieldas** [comment]
Here's the order I'm applying:

1. Provision with ansible, all apps are unconfigured. Buildarr config:

```
includes:
  - prowlarr.yml
#  - prowlarr.indexers.yml
#  - prowlarr.apps.yml
  - radarr.yml
  - sonarr.yml
 ```
2. Get a successful `buildarr run` 
3. Try another run with indexers applied:

```
includes:
  - prowlarr.yml
  - prowlarr.indexers.yml
#  - prowlarr.apps.yml
  - radarr.yml
  - sonarr.yml
 ```
5. That's successful so another run with apps applied:

```
includes:
  - prowlarr.yml
  - prowlarr.indexers.yml
  - prowlarr.apps.yml
  - radarr.yml
  - sonarr.yml
 ```

At this point I see `KeyError: 'syncRejectBlocklistedTorrentHashesWhileGrabbing'` 


prowlarr.apps.yml:

```
prowlarr:
  settings:
    apps:
      applications:
        delete_unmanaged: false
        definitions:
          Sonarr:
            type: sonarr
            instance_name: default  # The default Sonarr instance.
            base_url: http://localhost:8989
            prowlarr_url: http://localhost:9696
          Radarr:
            type: radarr
            instance_name: default  # The default Radarr instance.
            base_url: http://localhost:7878
            prowlarr_url: http://localhost:9696
      sync_profiles:
        delete_unmanaged: false
        definitions:
          Standard:
            enable_rss: true
            enable_interactive_search: true
            enable_automatic_search: true
            minimum_seeders: 1
```

**Comment by ezekieldas** [comment]
...but what I'm finding is even after this error, and as this error persists, my applications test fine and I see indexers on the associated applications.

**Comment by Callum027** [comment, marked as answer]
Hi @ezekieldas, thank you for the detailed troubleshooting information.

Looks like this is a new parameter for *Arr-type application definitions, added in [Prowlarr v1.15.0.4631](https://github.com/Prowlarr/Prowlarr/releases/tag/v1.15.0.4361).

Looks like this particular part of the Buildarr Prowlarr plugin is not tolerant of new fields it is not aware of. I'll make a PR to fix it and get it released as soon as possible.

**Reply by ezekieldas** [reply]
Hey thanks. Don't rush it on my part. Even with this error I just tested adding content to both Sonarr/Radarr to success. Unless it might be a blocker on consecutive modifications in buildarr, I can wait for a fix. 

Also, once I get more familiar with the codebase I might be able to help with some of these fixes as well. It's next level Python for me. Most of what I've written so far has been somewhat janky, glorified shell scripting.

**Reply by Callum027** [reply]
Created two issues (one for the immediate error, and another for the long term solution to issues like these Buildarr wide):

* https://github.com/buildarr/buildarr-prowlarr/issues/65
* https://github.com/buildarr/buildarr/issues/211

**Reply by Callum027** [reply]
Hey @ezekieldas, I have uploaded fixes to this PR: https://github.com/buildarr/buildarr-prowlarr/pull/66

I'll be doing testing on my own setup, but I'd appreciate it if you could test this on your setup as well. It's also a great way to get your feet wet when doing development in Buildarr (as a Python project, I'd classify this one as difficult to get into).

Here's a development `Dockerfile` that will install Buildarr, all plugins (including the Prowlarr plugin using the above PR). It can be run using the same configuration as the public Docker image (though it ignores `PUID` and `PGID`, it just runs as `root`).

Please let me know how you get on.

```dockerfile
# Dockerfile
# Buildarr Docker image.
#
# Copyright (C) 2023 Callum Dickinson
#
# Buildarr is free software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
#
# Buildarr is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with Buildarr.
# If not, see <https://www.gnu.org/licenses/>.

FROM python:3.11-alpine

# Mark the Buildarr configuration folder as an external volume mount point.
VOLUME ["/config"]

# Install Buildarr.
RUN apk add su-exec tzdata git && \
    python -m pip install --no-cache-dir "buildarr @ git+https://github.com/buildarr/buildarr.git@v0.7.1" \
                                         "buildarr-sonarr @ git+https://github.com/buildarr/buildarr-sonarr.git@v0.6.4" \
                                         "buildarr-radarr @ git+https://github.com/buildarr/buildarr-radarr.git@v0.2.6" \
                                         "buildarr-prowlarr[sonarr,radarr] @ git+https://github.com/buildarr/buildarr-prowlarr.git@65-add-sync_reject_blocklisted_torrent_hashes" \
                                         "buildarr-jellyseerr[sonarr,radarr] @ git+https://github.com/buildarr/buildarr-jellyseerr.git@v0.3.2"

# Set the Buildarr configuration folder as the default Docker container working folder.
WORKDIR /config

# Setup the Docker image entry point to bootstrap packages and start Buildarr daemon.
ENTRYPOINT ["buildarr"]
CMD ["daemon"]
```

**Reply by Callum027** [reply]
Finished the PR. I believe I have resolved this particular issue.

Please give the above Docker container a try, and let me know if you run into any additional issues. I'll add fixes for those to the PR.

**Reply by ezekieldas** [reply]
Incredible. Thanks so much for this. In short, this worked perfectly. 

Now, with my [modified Ansible Role](https://github.com/buildarr/ansible-playbooks), I'm now able to run this against a newly launched VM and a perfectly functional content pipeline becomes available. 

What I'm going to do now:

- Clean up the modified role and submit a final PR
- Take a closer look at your collection and scope modifications (possibly inspired from the modified role)
- Take a closer look at buildarr core and plugins and help with things like getting remote paths into Radarr!

**Reply by Callum027** [reply]
Thanks for confirming everything's working on your end.

I have made a new public releases with these fixes:

* Docker Container: [`callum027/buildarr:0.7.8`](https://hub.docker.com/r/callum027/buildarr) (or update your local container if using `latest`)
* Windows Installer: [`buildarr-installer-0.7.8-x64.exe`](https://github.com/buildarr/buildarr-installer/releases/tag/v0.7.8)
* Python Package: [`buildarr-prowlarr==0.5.3`](https://pypi.org/project/buildarr-prowlarr/0.5.3)
